lib/stages/VisSharedMemWriter.cpp

```diff
@@ -0,0 +1,390 @@
+    // write frame to ring buffer at time_ind and freq_ind
+
+    uint8_t *buf_write_pos = buf_addr + ((time_ind * nfreq + freq_ind) * frame_size);
+    int64_t *access_record_write_pos = access_record_addr + (time_ind * nfreq + freq_ind);
+
+    DEBUG("Writing ringbuffer to time_ind {} and freq_ind {}\n", time_ind, freq_ind);
+
+    // notify that time_ind and freq_ind are being written to, by setting that
+    // location to invalid in the access record
+    if (sem_wait(sem) == -1) {
+        FATAL_ERROR("Failed to acquire semaphore {}", _sem_name);
+        return;
+    }
+    memcpy(access_record_write_pos, &invalid, sizeof(invalid));
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+
+    // first write the metadata, then the data, then the valid byte
+    // add valid_size amount of padding
+    memcpy(buf_write_pos + valid_size, frame.metadata(), metadata_size);
+    memcpy(buf_write_pos + metadata_size + valid_size, frame.data(), data_size);
+    memcpy(buf_write_pos, &valid, sizeof(valid));
+
+    // Document the fpga sequence counter for that frame in the access record
+    uint64_t fpga_seq = frame.metadata()->fpga_seq_start;
+
+
+    if (sem_wait(sem) == -1) {
+        FATAL_ERROR("Failed to acquire semaphore {}", _sem_name);
+        return;
+    }
+
+    INFO("Writing fpga_seq {} to time index {}\n", fpga_seq, time_ind);
+    memcpy(access_record_write_pos, &fpga_seq, sizeof(fpga_seq));
+
+    // update num_writes
+    num_writes++;
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
+
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+
+}
+
+uint64_t VisSharedMemWriter::get_data_size(const visFrameView& frame) {
+
+    auto& dm = datasetManager::instance();
+
+    // Get properties of stream from first frame and datasetManager
+    // If we can get the data_size in other ways, we will not need
+    // ninput, nvis, or num_ev anymore
+    const stackState* stack_state = dm.dataset_state<stackState>(frame.dataset_id);
+    const inputState* input_state = dm.dataset_state<inputState>(frame.dataset_id);
+    const prodState* prod_state = dm.dataset_state<prodState>(frame.dataset_id);
+
+    const eigenvalueState* eigenvalue_state = dm.dataset_state<eigenvalueState>(frame.dataset_id);
+
+    if (!input_state || !prod_state) {
+        ERROR("Required datasetState not found for dataset ID {}\nThe following required states "
+                "were found:\ninputState - {:p}\nprodState - {:p}\n",
+                frame.dataset_id, (void*)input_state, (void*)prod_state);
+        throw std::runtime_error("Could not write to shared memory.");
+    }
+
+    // Count the eigenvalue index
+    size_t num_ev;
+    if (eigenvalue_state) {
+        num_ev = eigenvalue_state->get_num_ev();
+    } else {
+        num_ev = 0;
+    }
```

Comment:
Every one of the parameters you are looking up in here are exposed already on the `frame` object, so you can (and should) remove all the dM calls here.

Reply:
Can you review? https://github.com/kotekan/kotekan/pull/686/commits/0ff3351f19975d122de8d09b7e3fcd89efd26ed1